File: scripts/lib/materials-ingest/creator_ownership_write_v1.py
"""Materials Creator ownership WRITE — optional metadata.creator_user_id.

Mirrors scripts/lib/materials-ingest/creator-ownership-write-v1.mjs.
Platform/system generators omit the field unless an official TASFUL Account exists.
Never invents owners. Never trusts a client-supplied id alone.
"""

from __future__ import annotations

import re
from typing import Any

WRITE_CONTRACT_VERSION = "materials_creator_ownership_write.v1"

OWNER_FIELDS = (
    "creator_user_id",
    "seller_user_id",
    "sellerUserId",
    "owner_id",
    "user_id",
)

REJECT_EXACT = {
    "auto-supply",
    "autosupply",
    "platform",
    "official",
    "tasful",
    "tasful-official",
    "tasful_official",
    "operator",
    "unknown",
    "none",
    "null",
    "undefined",
    "n/a",
    "na",
    "anonymous",
    "system",
}

REJECT_SUBSTRING = (
    "comfyui",
    "auto-generator",
    "autogenerator",
    "local-synthesis",
    "ace-step",
    "template-local",
    "presentation-local",
    "generator",
)

ACCOUNT_ID_RE = re.compile(r"^[A-Za-z0-9._:-]+$")
PLATFORM_SOURCES = {"platform", "generator", "system", "auto-supply", "autosupply"}
USER_SOURCES = {"user", "creator"}


def strip_owner_fields(meta: dict[str, Any] | None) -> dict[str, Any]:
    out = dict(meta or {})
    for field in OWNER_FIELDS:
        out.pop(field, None)
    out.pop("__ownership", None)
    return out


def _looks_like_account_user_id(value: str) -> bool:
    if len(value) < 2 or len(value) > 80:
        return False
    if "@" in value or "/" in value or "\\" in value or "://" in value:
        return False
    return bool(ACCOUNT_ID_RE.match(value))


def _resolve_canonical(value: str) -> tuple[bool, str]:
    raw = str(value or "").strip()
    if not raw:
        return False, "unresolved"
    lowered = raw.lower()
    if lowered in REJECT_EXACT:
        return False, "rejected_token"
    if any(token in lowered for token in REJECT_SUBSTRING):
        return False, "rejected_generator_or_provider"
    if not _looks_like_account_user_id(raw):
        return False, "not_tasful_account_user_id"
    return True, raw


def apply_creator_ownership(
    meta: dict[str, Any] | None,
    *,
    source: str = "platform",
    authenticated_user_id: str | None = None,
    claimed_user_id: str | None = None,
    platform_creator_user_id: str | None = None,
) -> dict[str, Any]:
    cleaned = strip_owner_fields(meta)
    src = str(source or "platform").strip().lower()
    claimed = str(claimed_user_id or "").strip()

    if src in PLATFORM_SOURCES:
        official = str(platform_creator_user_id or "").strip()
        if not official:
            return cleaned
        ok, canonical = _resolve_canonical(official)
        if not ok:
            return cleaned
        if claimed and claimed != canonical:
            raise ValueError("client_spoof_denied")
        cleaned["creator_user_id"] = canonical
        return cleaned

    if src not in USER_SOURCES:
        raise ValueError("unknown_ownership_source")

    authenticated = str(authenticated_user_id or "").strip()
    if not authenticated:
        raise ValueError("authenticated_user_required")
    ok, canonical = _resolve_canonical(authenticated)
    if not ok:
        raise ValueError("authenticated_user_not_canonical")
    if claimed and claimed != canonical:
        raise ValueError("client_spoof_denied")
    cleaned["creator_user_id"] = canonical
    return cleaned


def apply_platform_generator_metadata(meta: dict[str, Any] | None) -> dict[str, Any]:
    return apply_creator_ownership(meta, source="platform")

File: scripts/lib/materials-ingest/creator-ownership-write-v1.mjs
/**
 * Materials Creator ownership WRITE — optional metadata.creator_user_id.
 * Reuses resolveCanonicalCreatorUserId. Does not invent owners.
 * Client-supplied creator_user_id is never trusted alone.
 */
import {
  CANONICAL_OWNER_ID_FIELDS,
  resolveCanonicalCreatorUserId,
} from "./creator-ownership-v1.mjs";

export const WRITE_CONTRACT_VERSION = "materials_creator_ownership_write.v1";

export const OWNERSHIP_SOURCE = Object.freeze({
  USER: "user",
  CREATOR: "creator",
  PLATFORM: "platform",
  GENERATOR: "generator",
  SYSTEM: "system",
});

/** Existing TASFUL Account demo identity used by Creator Page / listing-seller-profile. */
export const CANONICAL_QA_USER_ID = "u_hiro";

const PLATFORM_SOURCES = new Set([
  OWNERSHIP_SOURCE.PLATFORM,
  OWNERSHIP_SOURCE.GENERATOR,
  OWNERSHIP_SOURCE.SYSTEM,
  "auto-supply",
  "autosupply",
]);

const USER_SOURCES = new Set([OWNERSHIP_SOURCE.USER, OWNERSHIP_SOURCE.CREATOR]);

export class CreatorOwnershipWriteError extends Error {
  constructor(result = {}) {
    super(result.reason || "creator_ownership_denied");
    this.name = "CreatorOwnershipWriteError";
    this.code = result.reason || "creator_ownership_denied";
    this.result = result;
  }
}

function firstClaimedOwnerId(source = {}) {
  if (!source || typeof source !== "object") return "";
  for (const field of CANONICAL_OWNER_ID_FIELDS) {
    const value = String(source[field] == null ? "" : source[field]).trim();
    if (value) return value;
  }
  return "";
}

export function stripOwnerFields(meta = {}) {
  const out = { ...(meta && typeof meta === "object" ? meta : {}) };
  for (const field of CANONICAL_OWNER_ID_FIELDS) {
    delete out[field];
  }
  delete out.__ownership;
  return out;
}

/**
 * Trusted write decision. Never writes generator / platform tokens.
 * @param {{
 *   source?: string,
 *   authenticatedUserId?: string,
 *   claimedUserId?: string,
 *   platformCreatorUserId?: string,
 * }} input
 */
export function resolveTrustedWriteCreatorUserId(input = {}) {
  const source = String(input.source || OWNERSHIP_SOURCE.PLATFORM).trim().toLowerCase();
  const claimed = String(input.claimedUserId == null ? "" : input.claimedUserId).trim();

  if (PLATFORM_SOURCES.has(source)) {
    const official = String(input.platformCreatorUserId == null ? "" : input.platformCreatorUserId).trim();
    if (!official) {
      return {
        ok: true,
        omit: true,
        reason: "platform_no_official_identity",
      };
    }
    const resolved = resolveCanonicalCreatorUserId({ creator_user_id: official });
    if (!resolved.ok) {
      return {
        ok: true,
        omit: true,
        reason: "platform_identity_not_canonical",
        detail: resolved.reason,
      };
    }
    if (claimed && claimed !== resolved.creator_user_id) {
      return { ok: false, reason: "client_spoof_denied", claimed, trusted: resolved.creator_user_id };
    }
    return {
      ok: true,
      omit: false,
      creator_user_id: resolved.creator_user_id,
      source_field: "platformCreatorUserId",
    };
  }

  if (!USER_SOURCES.has(source)) {
    return { ok: false, reason: "unknown_ownership_source", source };
  }

  const authenticated = String(input.authenticatedUserId == null ? "" : input.authenticatedUserId).trim();
  if (!authenticated) {
    return { ok: false, reason: "authenticated_user_required" };
  }
  const trusted = resolveCanonicalCreatorUserId({ creator_user_id: authenticated });
  if (!trusted.ok) {
    return { ok: false, reason: "authenticated_user_not_canonical", detail: trusted.reason };
  }
  if (claimed && claimed !== trusted.creator_user_id) {
    return {
      ok: false,
      reason: "client_spoof_denied",
      claimed,
      trusted: trusted.creator_user_id,
    };
  }
  return {
    ok: true,
    omit: false,
    creator_user_id: trusted.creator_user_id,
    source_field: "authenticatedUserId",
  };
}

/**
 * Apply optional creator_user_id. Fail-closed on spoof.
 * @returns {{ ok: true, metadata: object, omitted?: boolean, creator_user_id?: string } | { ok: false, reason: string, metadata: object }}
 */
export function applyCreatorOwnershipToMetadata(meta = {}, ctx = {}) {
  const cleaned = stripOwnerFields(meta);
  const decided = resolveTrustedWriteCreatorUserId({
    source: ctx.source,
    authenticatedUserId: ctx.authenticatedUserId,
    claimedUserId: ctx.claimedUserId || firstClaimedOwnerId(meta),
    platformCreatorUserId: ctx.platformCreatorUserId,
  });
  if (!decided.ok) {
    return { ...decided, metadata: cleaned };
  }
  if (decided.omit) {
    return { ok: true, omitted: true, reason: decided.reason, metadata: cleaned };
  }
  return {
    ok: true,
    omitted: false,
    creator_user_id: decided.creator_user_id,
    metadata: { ...cleaned, creator_user_id: decided.creator_user_id },
  };
}

/** Platform/generator writers: strip owner fields, never invent. */
export function applyPlatformGeneratorMetadata(meta = {}) {
  const applied = applyCreatorOwnershipToMetadata(meta, { source: OWNERSHIP_SOURCE.PLATFORM });
  return applied.metadata;
}

/**
 * Existing-register: keep a canonical owner already on disk evidence; never invent.
 */
export function preserveTrustedPackageCreatorUserId(meta = {}) {
  const resolved = resolveCanonicalCreatorUserId(meta);
  const cleaned = stripOwnerFields(meta);
  if (resolved.ok) {
    cleaned.creator_user_id = resolved.creator_user_id;
  }
  return cleaned;
}

export function extractOwnershipContext(extraMetadata = {}) {
  const extra = extraMetadata && typeof extraMetadata === "object" ? { ...extraMetadata } : {};
  const ownership = extra.__ownership && typeof extra.__ownership === "object" ? extra.__ownership : {};
  const claimedUserId = firstClaimedOwnerId(extra);
  delete extra.__ownership;
  for (const field of CANONICAL_OWNER_ID_FIELDS) {
    delete extra[field];
  }
  return {
    extra,
    ownership,
    claimedUserId,
    source: ownership.source || OWNERSHIP_SOURCE.PLATFORM,
    authenticatedUserId: ownership.authenticatedUserId,
    platformCreatorUserId: ownership.platformCreatorUserId,
  };
}

export function applyOwnershipOrThrow(meta, ctx) {
  const applied = applyCreatorOwnershipToMetadata(meta, ctx);
  if (!applied.ok) {
    throw new CreatorOwnershipWriteError(applied);
  }
  return applied.metadata;
}
